src/Enzo/enzo_EnzoMethodStarMakerSTARSS.cpp

```diff
@@ -0,0 +1,567 @@
+        
+     
+        // compute MMW -- TODO: Make EnzoComputeMeanMolecularWeight class and reference
+        // mu_field here?
+        if (primordial_chemistry == 0) mu = enzo_config->ppm_mol_weight;
+        else {
+          mu = d_el[i] + dHI[i] + dHII[i] + 0.25*(dHeI[i]+dHeII[i]+dHeIII[i]);
+
+          if (primordial_chemistry > 1) {
+            mu += dHM[i] + 0.5*(dH2I[i]+dH2II[i]);
+          }
+          if (primordial_chemistry > 2) {
+            mu += 0.5*(dDI[i] + dDII[i]) + dHDI[i]/3.0;
+          }
+        }
+        mu /= density[i]; 
+        mu = 1.0/mu;
+    
+
+        double rho_cgs = density[i] * enzo_units->density();
+        double mean_particle_mass = mu * cello::mass_hydrogen;
+        double ndens = rho_cgs / mean_particle_mass;
+
+        double cell_mass  = density[i] * cell_volume;
+        double metallicity = (metal) ? metal[i]/density[i]/cello::metallicity_solar : 0.0;
+
+        //
+        // Apply the criteria for star formation
+        //
+        
+        // check number density threshold
+        if (! this->check_number_density_threshold(ndens) ) continue;
+
+        // check overdensity threshold
+        // In cosmology, units are scaled such that mean(density) = 1,
+        // so density IS overdensity in these units        
+        if (! this->check_overdensity_threshold(density[i]) ) continue;
+
+        #ifdef DEBUG_SF_CRITERIA 
+           CkPrintf("MethodFeedbackSTARSS -- overdensity threshold passed! rho=%f\n",dmean);
+        #endif
+        
+        // check velocity divergence < 0
+        if (! this->check_velocity_divergence(velocity_x, velocity_y,
+                                              velocity_z, i,
+                                              idx, idy, idz)) continue;
+      
+        #ifdef DEBUG_SF_CRITERIA_EXTRA
+           CkPrintf("MethodStarMakerSTARSS -- div(v) < 0 in cell %d\n", i);
+        #endif 
+
+        // check that alpha < 1
+        if (use_altAlpha_) {
+          // NOTE: this requires the DEBUG_COPY_POTENTIAL flag to be set in EnzoMethodGravity
+          // TODO: Make saving potential an input parameter instead of debug flag?
+          if (! this->check_self_gravitating_new(total_energy[i], potential[i])) continue;
+        }
+
+        else {
+          if (! this->check_self_gravitating(mean_particle_mass, density[i], temperature[i],
+                                             velocity_x, velocity_y, velocity_z,
+                                             lunit, vunit, rhounit,
+                                             i, idx, idy, idz, dx, dy, dz)) continue;
+        }
+
+        #ifdef DEBUG_SF_CRITERIA_EXTRA
+           CkPrintf("MethodStarMakerSTARSS -- alpha < 1 in cell %d\n", i);
+        #endif 
+
+        // check that (T<Tcrit) or (dynamical_time < cooling_time)
+        // In order to check cooling time, must have use_temperature_threshold=true;
+        double total_density = density[i] + density_particle_accumulate[i];
```

Comment:
I'm fairly sure that "density_particle_accumulate" is set to zero at the end of the gravity method. To be honest, I'm not sure "density_particle" and "density_particle_accumulate" are even used at all in the gravity method.

I think a better idea would be to have a "density_total_copy" field, which copies the "density_total" field in the gravity method. I was thinking of doing something similar in my "sink-formation" branch (https://github.com/stefanarridge/enzo-e/blob/sink-formation). You could then put in a requirement that the gravity method precedes this method (using `cello::problem()->method_precedes`).